src/Stratis.Bitcoin.Features.Consensus.Tests/TimeSyncBehaviorAdjustedTimeTest.cs

```diff
@@ -0,0 +1,96 @@
+        /// uses outbound samples as a priority over inbound as they are less likely to be malicious.
+        /// This means that even when there are many inbound connections, only the first ones are considered until there are enough outbound nodes. 
+        /// Eg only 9 inbound per 1 outbound.
+        /// </summary>
+        [Fact]
+        public void AddTimeData_WithLargeSampleSetOfInboundTimeManipulatorsAndLowSampleSetOfOutbound_GetsOverridenByOutboundSamples()
+        {
+            this.Given_an_empty_time_sync_behaviour_state();
+            this.Given_40_inbound_samples_with_offset_of(10);
+            this.Given_1_outbound_sample_with_offset_of(20);
+            this.When_calculating_time_adjust_offset();
+            this.Then_adjusted_time_offset_should_be(20);
+        }
+
+        /// <summary>
+        /// Checks that <see cref="TimeSyncBehaviorState.AddTimeData(IPAddress, TimeSpan, bool)"/>
+        /// uses system time without an offset when there are no outbound samples.
+        /// </summary>
+        [Fact]
+        public void AddTimeData_WithLargeSampleSetOfInboundTimeManipulatorsAndZeroOutbound_SticksWithTheSystemTime()
+        {
+            this.Given_an_empty_time_sync_behaviour_state();
+            this.Given_40_inbound_samples_with_offset_of(10);
+            this.When_calculating_time_adjust_offset();
+            this.Then_adjusted_time_offset_should_be(0);
+        }
+
+        private void Given_an_empty_time_sync_behaviour_state()
+        {
+            this.dateTimeProvider = new DateTimeProvider();
+            this.timesyncBehaviourState = new TimeSyncBehaviorState(
+                this.dateTimeProvider, new NodeLifetime(),
+                new AsyncLoopFactory(new LoggerFactory()), 
+                new LoggerFactory());
+        }
+
+        private void Given_1_outbound_sample_with_offset_of(int offsetSeconds)
+        {
+            this.timesyncBehaviourState.AddTimeData(IPAddress.Parse("2.2.2.2"), TimeSpan.FromSeconds(offsetSeconds), isInboundConnection: false);
+        }
+
+        private void Given_40_inbound_samples_with_offset_of(int offset)
+        {
+            for (int i = 1; i <= 40; i++)
+            {
+                this.timesyncBehaviourState.AddTimeData(IPAddress.Parse("1.2.3." + i), TimeSpan.FromSeconds(offset), isInboundConnection: true);
+            }
+        }
+
+        private void When_calculating_time_adjust_offset()
+        {
+            var adjustedTime = this.dateTimeProvider.GetAdjustedTime();
+            var now = this.dateTimeProvider.GetUtcNow();
+            this.offset = adjustedTime - now;
+        }
+
+        private void Then_adjusted_time_offset_should_be(int expectedOffset)
+        {
+            /// It is only rounded so that assertions are easier, a future improvement would be to not use UtcNow directly in the code but to use an indirection, allowing it to be replaced at test time.
```

Comment:
Actually I will remove the part about future improvement as that doesn't need to live in the code. It is an overall concept that can be applied for testing DateTimes effectively.  But since implemented using rounding here, there is no real value in me leaving that extra part in.

I would rather not have a comment at all